emission/exportdata/export_data.py

```diff
@@ -0,0 +1,50 @@
+# Standard imports 
+import logging
+
+# Our imports
+import emission.storage.pipeline_queries as epq
+import emission.storage.timeseries.abstract_timeseries as esta
+import emission.storage.decorations.analysis_timeseries_queries as esda
+
+import gzip
+import json
+import bson.json_util as bju
+
+def set_export_data(user_id):
+    time_query = epq.get_time_range_for_export_data(user_id)        # Does this have .startTs and .endTs atributes? Should we use those?
+    try: 
+        edp = ExportDataPipeline()
+        edp.user_id = user_id
+        edp.run_export_data_pipeline(user_id, time_query)
+        if edp.last_trip_done is None:
+            logging.debug("After run, last_trip_done == None, must be early return")
+        epq.mark_export_data_done(user_id, edp.last_trip_done)
+    except:
+        logging.exception("Error while exporting, timestamp unchanged")
+        epq.mark_export_data_failed(user_id)
+
+class ExportDataPipeline:
+    def __init__(self):
+        self._last_trip_done = None
+
+    @property
+    def last_trip_done(self):
+        return self._last_trip_done
+
+    def run_export_data_pipeline(self, user_id, time_query):
+        ts = esta.TimeSeries.get_time_series(user_id)
+        
+        #This will essentially give us a list of the raw data in return. There are no restrictions on what is added, all keys are added so long as they fall in the time range.
+        raw_data_list = list(ts.find_entries(key_list=None, time_query=time_query))
+
+        #This will give us just a list of the cleaned trips. (Another example option)
+        #cleaned_trip_list = list(ts.find_entries(key_list = {'analysis/cleaned_trip'}, time_query=time_query))
+
+        #File export
+        export_file = "%s_%s.gz" % ("export", user_id)
+        
+        with gzip.open(export_file, "wt") as ef:
+            json.dump(raw_data_list, ef, default=bju.default, allow_nan=False, indent=4)
```

Comment:
This works for the current one-time export from the script because we are only exporting data once per user. But with the incremental updates, we may be exporting data every hour. With this file naming, we will overwrite the file every time, which is not very robust.

We really need to discuss with @kafitz on what the plans for pulling the data from this stage are.

Reply:
@kafitz What would be the best strategy to combat this? As Shankari mentioned, we run into the issue of overwriting the file each time. Is there a better naming strategy, or can you time the pulling of the data to avoid this?

Reply:
@kafitz alternatively, if we want to write and expose a small REST API, then we never have to write to a file.

BUT we do have to deal with a separate stage to ensure that the data was transmitted correctly; what if @kafitz's code invokes the API, we generate the result, we update the pipeline stage, and then partway through reading the data, the receiving code dies (maybe OOM).

The next call would then retrieve the next batch of data, and some data would be lost.

If we want to write to shared file storage, then we won't have that problem, but we *will* have the problem of naming the files accordingly.

@kafitz, which transfer mechanism do you prefer?

Reply:
Sorry I'm back now and getting re-acclimated. An issue with writing chunked exports to unique files is we then create a directory of mostly useless data (after it's been downloaded once) that will need to be cleaned up. Therefore, I think it would be cleanest to go the REST API route.

My other assumption (that could be wrong) is that I expect eventually the data may grow large enough that calling a REST API to try to export, bundle and send all data may timeout. Therefore I was conceptually some sort of long-polling or socket approach--as long as it doesn't overcomplicate.

I'd be more in a favor of an API that worked either by:
1. Client REST request to server to begin the export. Server begins export to a common filename that can be overwritten each time and tracks current export state. Client polls to check if export is available and downloads the static .gz file when it receives a state of OK. There could be a final API call here to clean up the received file from the server.
2. Client opens 2-way socket that will encapsulate the full export process. Client signals to server to begin export, server performs export and signals back to client that export is ready, exported data is then streamed to the client to handle however it wants.

With either approach, I was envisioning that I would initially send a basic GET/POST request with the ts of the last received data for each user--does that sound correct?